### DISCUSSION TITLE
不统一的DOM元素指定方式

### DISCUSSION THREAD

**Discussion by mouse0w0** [layui/layui#2227] [讨论]
在使用layui需要指定某一DOM元素时，指定的方式并不统一使我深受困扰，以下是我遇到的一些问题：

1. 渲染表单时，使用`form.render($('#form-demo'))`，不能使用`form.render('#form-demo')`以避免使用JQuery。

2. 获取表单数据时，使用lay-filter选择`form.val('form-demo-filter')`，不能使用ID选择器`form.val('#form-demo')`或者JQuery`form.val($('#form-demo'))`。

3. 重载表格时，使用table自己的id选择`table.reload('table-demo-id')`，不能使用ID选择器`table.reload('#table-demo')`或者JQuery。

此外还有一些指定模板DOM元素的问题，比如table的toolbar属性不支持function，layer的content属性不支持ID选择器。

以上是我在使用layui中遇到的问题，但仍有很多类似问题未能一一列举，如有不正确的地方欢迎各位指正，希望layui能够改进这些问题。

**Comment by sentsim** [comment]
1. `$('#form-demo')` 和 `'#form-demo'` 是不同的数据类型，支持传入 jQuery 对象进行定向渲染，是 2.7 开始新增的，在此之前，`form.render()` 的第一个参数接受的是 string 类型，即指定渲染的是什么表单类型，如 `input,select,checkbox,radio`。若将 `$('#form-demo')` 和 `'#form-demo'`  视为等同，将不得不写更多兼容的代码，事实上这完全没必要，文档已经清晰地说明当它接受的是 jQuery 对象时，即为定向渲染。
2. `form.val()` 在文档明确指出接受的是 table 元素的 lay-filter 属性值。只是后面为了方便，同时也兼容了 table 实例 ID，那么既然是 ID，就没必要加个 `#`，那样就是选择器了。好比 `document.getElementById('ID')`，而不是 `document.getElementById('#ID')`
3. 跟第二点问题类似。

**Layui 的可接受参数的多样性，是版本迭代过程中功能不断演进又不得不向下兼容的结果**。能尽量形成统一当然是最好的，但你提到的几点，其实并不在统一标准的范畴之内。**当我们使用一个主要面向业务层的 UI 组件库时，往往很难做到完全的随心所欲，该遵循文档的还是要遵循**。而你的这些困扰，也恰恰是渴望随心所欲的心理落差。

**Comment by mouse0w0** [comment]
有时用lay-filter，有时用JQuery，有时要写#，有时不写#，这些不统一的元素指定给layui的使用者（比如我）造成了不小的心智负担。可能现在为了向下兼容不得不这样做，但我个人认为应该让使用者自然的使用CSS选择器来指定元素，而不是总是在关注不同的函数如何输入参数来指定元素，例如`form.render()`和`form.val()`。

**Reply by sentsim** [reply]
`lay-filter` 在 table 中其实已被 `id` 逐渐淘汰。
